pkg/controller/deployment/deployment_controller.go

```diff
@@ -438,41 +438,150 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+			return
+		}
+		if newRS.Annotations == nil {
+			newRS.Annotations = make(map[string]string)
+		}
+		newRS.Annotations[deploymentutil.DesiredReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+		newRS.Annotations[deploymentutil.TotalReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+		glog.V(4).Infof("Scaling new replica set %q from %d to %d", newRS.Name, newRS.Spec.Replicas, deployment.Spec.Replicas)
+		if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+			glog.Infof("Cannot scale replica set %q to %d replicas: %v", newRS.Name, deployment.Spec.Replicas, err)
+		}
+		return
+	}
+
+	// If the new replica set is saturated, old replica sets should be fully scaled down.
+	// This case handles replica set adoption during a saturated new replica set.
+	if deploymentutil.IsSaturated(deployment, newRS) {
+		for _, old := range controller.FilterActiveReplicaSets(oldRSs) {
+			if old.Annotations == nil {
+				old.Annotations = make(map[string]string)
+			}
+			old.Annotations[deploymentutil.DesiredReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+			old.Annotations[deploymentutil.TotalReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+			glog.V(4).Infof("Scaling down replica set %q from %d to 0", old.Name, old.Spec.Replicas)
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(old, 0, deployment); err != nil {
+				glog.Infof("Cannot scale replica set %q to 0 replicas: %v", old.Name, err)
+			}
+		}
+		return
+	}
+
+	// There are old replica sets with pods and the new replica set is not saturated.
+	// We need to proportionally scale all replica sets (new and old) in case of a
+	// rolling deployment.
+	if deploymentutil.IsRollingUpdate(deployment) {
+		sort.Sort(controller.BySize(allRSs))
+
+		totalReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+		totalAdded := 0
+		// Number of additional pods that can be either added or removed from the total pod count.
+		additional := deployment.Spec.Replicas - totalReplicas
+		if additional == 0 {
+			return
+		}
+
+		for i := range allRSs {
+			rs := allRSs[i]
+			total := totalReplicas
+			// Use the deployment.kubernetes.io/total-replicas annotation to determine the proportion
+			// for this deployment. If we cannot find the annotation then the current total amount
+			// will be used.
+			totalDesiredString, ok := rs.Annotations[deploymentutil.TotalReplicasAnnotation]
+			totalDesired, err := strconv.Atoi(totalDesiredString)
+			if ok && err == nil {
+				total = int32(totalDesired)
+			}
+			proportion := 0
+			if additional > 0 {
+				// Scaling up
+				proportion = integer.IntMin(integer.RoundToInt((float64(rs.Spec.Replicas)/float64(total))*float64(additional)), int(additional)-totalAdded)
+			} else {
+				// Scaling down
+				proportion = integer.IntMax(integer.RoundToInt((float64(rs.Spec.Replicas)/float64(total))*float64(additional)), int(additional)-totalAdded)
```

Comment:
It would help code clarity to break this calculation out into a couple functions and intermediate variables.

e.g.
GetFractionOfDeploymentReplicas(rs) float64 // Returns the fraction of desired replicas in the deployment contained in this replica set.  (note, desired replicas does not necessarily equal actual replicas due to max surge and unavailable)
GetAdditionalReplicas(rs, totalToAdd, remainingToAdd) int // Returns the number of replicas to be added to this rs when scaling the deployment by totalToAdd cnt.